File: src/main/java/com/ab/ploy/service/WordService.java
/* Copyright © 2021 */
package com.ab.ploy.service;

import com.ab.ploy.model.Language;
import com.ab.ploy.model.SimpleWord;
import com.ab.ploy.model.TranslatedLanguage;
import com.ab.ploy.model.Word;
import com.ab.ploy.persistence.WordRepository;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.springframework.stereotype.Service;

@Service
public class WordService {

  private final WordRepository wordRepository;

  public WordService(WordRepository wordRepository) {
    this.wordRepository = wordRepository;
  }

  public List<Word> getWords() {
    return wordRepository.findAll();
  }

  public List<Word> createWord(Word word) {
    Word savedWord = wordRepository.save(word);

    List<Word> propagatedWords = propagate(word);

    List<Word> allSavedWords = new ArrayList<>();
    allSavedWords.add(savedWord);
    allSavedWords.addAll(propagatedWords);

    return allSavedWords;
  }

  private List<Word> propagate(Word word) {
    List<Word> result = new ArrayList<>();

    for (Language language : Language.values()) {
      Word propagatedWord =
          propagateLanguage(word, word.getTranslatedLanguages().get(language), language);
      if (propagatedWord != null) {
        result.add(word);
      }
    }

    return result;
  }

  private Word propagateLanguage(
      Word word, TranslatedLanguage translatedLanguage, Language language) {
    if (translatedLanguage == null) return null;
    if (translatedLanguage.getPropagate() != Boolean.TRUE) return null;
    if (word.getLanguage() == language) return null;

    Map<Language, TranslatedLanguage> translatedLanguages = new HashMap<>();
    translatedLanguages.put(
        word.getLanguage(), new TranslatedLanguage(List.of(new SimpleWord(word.getWord()))));

    word.getTranslatedLanguages().keySet().stream()
        .filter(l -> l != language)
        .forEach(
            l ->
                translatedLanguages.put(
                    l,
                    new TranslatedLanguage(
                        List.of(
                            new SimpleWord(
                                word.getTranslatedLanguages()
                                    .get(l)
                                    .getTranslations()
                                    .get(0)
                                    .getWord())))));

    return new Word(
        null,
        language,
        translatedLanguage.getTranslations().get(0).getWord(),
        word.getType(),
        translatedLanguages);
  }
}

File: src/main/kotlin/com/ab/ploy/admin/service/WordService.kt
/* Copyright © 2021 */
package com.ab.ploy.admin.service

import com.ab.ploy.admin.persistence.WordRepository
import com.ab.ploy.common.models.Language
import com.ab.ploy.common.models.SimpleWord
import com.ab.ploy.common.models.TranslatedLanguage
import com.ab.ploy.common.models.Word
import org.springframework.stereotype.Service

@Service
class WordService(private val wordRepository: WordRepository) {


    fun getWord(word: String) = wordRepository.findByWord(word)

    fun getWords(): MutableList<Word> = wordRepository.findAll()

    fun createWord(word: Word): Map<Language, Int> {
        val savedWord = wordRepository.save(word)

        val propagatedWords = propagate(word)

        val allSavedWords = mutableListOf(savedWord)

        propagatedWords.forEach { allSavedWords.add(wordRepository.save(it)) }

        return allSavedWords.map { it.language to 1 }.toMap()
    }

    fun propagate(word: Word): List<Word> {
        val swedish =
            propagateLanguage(
                word, word.translatedLanguages?.get(Language.SWEDISH), Language.SWEDISH)
        val english =
            propagateLanguage(
                word, word.translatedLanguages?.get(Language.ENGLISH), Language.ENGLISH)
        val spanish =
            propagateLanguage(
                word, word.translatedLanguages?.get(Language.SPANISH), Language.SPANISH)

        return listOfNotNull(swedish, english, spanish)
    }

    fun propagateLanguage(
        word: Word,
        translatedLanguage: TranslatedLanguage?,
        language: Language
    ): Word? {
        if (translatedLanguage == null) return null
        if (translatedLanguage.propagate != true) return null
        if (word.language == language) return null // Don't propagate to the same language

        val translatedLanguages =
            mutableMapOf(Pair(word.language, TranslatedLanguage(listOf(SimpleWord(word.word)))))

        word.translatedLanguages?.filterKeys { it != language }?.forEach {
            translatedLanguages[it.key] =
                TranslatedLanguage(listOf(SimpleWord(it.value.translations!![0].word)))
        }

        return Word(
            null,
            language,
            translatedLanguage.translations?.get(0)?.word!!,
            word.type,
            translatedLanguages)
    }
}
